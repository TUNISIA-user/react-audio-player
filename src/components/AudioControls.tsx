import playSVG from "../icons/play.svg";
import pauseSVG from "../icons/pause.svg";
import skipSVG from "../icons/skip.svg";
import { type ReactElement } from "react";

interface Props {
  isPlaying: boolean;
  allowSkip: boolean;
  onPlayClick: () => void;
  onPauseClick: () => void;
  onSkipForwardClick: () => void;
  onSkipBackwardClick: () => void;
}

const AudioControls = ({
  isPlaying,
  allowSkip,
  onPlayClick,
  onPauseClick,
  onSkipForwardClick,
  onSkipBackwardClick,
}: Props): ReactElement => {
  return (
    <div
      style={{
        display: "flex",
        alignItems: "center",
        gridColumn: "2 / 2",
        gap: 15,
      }}
    >
      {/* skip backwards */}
      {allowSkip && (
        <img
          src={skipSVG}
          onClick={onSkipBackwardClick}
          title={"Skip backwards"}
          style={{
            cursor: "pointer",
            height: 16,
            width: 16,
            transform: "rotate(180deg)",
          }}
        />
      )}
      {/* skip fowards */}
      <img
        src={isPlaying ? pauseSVG : playSVG}
        onClick={isPlaying ? onPauseClick : onPlayClick}
        title={isPlaying ? "Pause" : "Play"}
        style={{
          cursor: "pointer",
          height: "20px",
          width: "20px",
        }}
      />
      {/* skip forwards */}
      {allowSkip && (
        <img
          src={skipSVG}
          onClick={onSkipForwardClick}
          title={"Skip forwards"}
          style={{
            cursor: "pointer",
            height: 16,
            width: 16,
          }}
        />
      )}
    </div>
  );
};

export default AudioControls;
