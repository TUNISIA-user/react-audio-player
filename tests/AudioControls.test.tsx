import React from "react";
import { expect, test, vi } from "vitest";
import renderer from "react-test-renderer"
import userEvent from "@testing-library/user-event";
import { render, act, screen } from "@testing-library/react";
import AudioControls from "../src/components/AudioControls";

describe("Test audio controls", () => {
  const onPlayClick = vi.fn()
  const onPauseClick = vi.fn()
  const onSkipForwardClick = vi.fn()
  const onSkipBackwardClick = vi.fn()

  test("AudioControls snapshot (playing + skip)", () => {
    const audioControls = renderer.create(
      <AudioControls 
        isPlaying={true}
        allowSkip={true}
        onPlayClick={onPlayClick}
        onPauseClick={onPauseClick}
        onSkipForwardClick={onSkipForwardClick}
        onSkipBackwardClick={onSkipBackwardClick}
      />
    )
    const tree = audioControls.toJSON();
    expect(tree).toMatchSnapshot();
  })

  test("AudioControls snapshot (no playing + no skip)", () => {
    const audioControls = renderer.create(
      <AudioControls 
        isPlaying={false}
        allowSkip={false}
        onPlayClick={onPlayClick}
        onPauseClick={onPauseClick}
        onSkipForwardClick={onSkipForwardClick}
        onSkipBackwardClick={onSkipBackwardClick}
      />
    )
    
    const tree = audioControls.toJSON();
    expect(tree).toMatchSnapshot();
  })

  test("AudioControls onPlayClick when paused", async () => {
    render(
      <AudioControls 
        isPlaying={false}
        allowSkip={false}
        onPlayClick={onPlayClick}
        onPauseClick={onPauseClick}
        onSkipForwardClick={onSkipForwardClick}
        onSkipBackwardClick={onSkipBackwardClick}
      />
    )

    const element = screen.getByTestId("play-pause")
    await act(() => userEvent.click(element))
    expect(onPlayClick).toHaveBeenCalledOnce();
  })

  test("AudioControls onPlayClick when playing", async () => {
    render(
      <AudioControls 
        isPlaying={true}
        allowSkip={false}
        onPlayClick={onPlayClick}
        onPauseClick={onPauseClick}
        onSkipForwardClick={onSkipForwardClick}
        onSkipBackwardClick={onSkipBackwardClick}
      />
    )

    const element = screen.getByTestId("play-pause")
    await act(() => userEvent.click(element))
    expect(onPauseClick).toHaveBeenCalledOnce();
  })

  test("AudioControls skips buttons", async () => {
    render(
      <AudioControls 
        isPlaying={true}
        allowSkip={true}
        onPlayClick={onPlayClick}
        onPauseClick={onPauseClick}
        onSkipForwardClick={onSkipForwardClick}
        onSkipBackwardClick={onSkipBackwardClick}
      />
    )

    const elementBack = screen.getByTestId("skip-back")
    await act(() => userEvent.click(elementBack))
    expect(onSkipBackwardClick).toHaveBeenCalledOnce();

    const elementForward = screen.getByTestId("skip-forward")
    await act(() => userEvent.click(elementForward))
    expect(onSkipForwardClick).toHaveBeenCalledOnce();
  })
})
