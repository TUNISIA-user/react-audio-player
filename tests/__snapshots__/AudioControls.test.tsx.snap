// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test trackbar > AudioControls snapshot (no playing + no skip) 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "gap": 15,
      "gridColumn": "2 / 2",
    }
  }
>
  <img
    data-testid="play-pause"
    onClick={[MockFunction spy]}
    src="/src/icons/play.svg"
    style={
      {
        "cursor": "pointer",
        "height": "20px",
        "width": "20px",
      }
    }
    title="Play"
  />
</div>
`;

exports[`Test trackbar > AudioControls snapshot (playing + skip) 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "gap": 15,
      "gridColumn": "2 / 2",
    }
  }
>
  <img
    data-testid="skip-back"
    onClick={[MockFunction spy]}
    src="/src/icons/skip.svg"
    style={
      {
        "cursor": "pointer",
        "height": 16,
        "transform": "rotate(180deg)",
        "width": 16,
      }
    }
    title="Skip backwards"
  />
  <img
    data-testid="play-pause"
    onClick={[MockFunction spy]}
    src="/src/icons/pause.svg"
    style={
      {
        "cursor": "pointer",
        "height": "20px",
        "width": "20px",
      }
    }
    title="Pause"
  />
  <img
    data-testid="skip-forward"
    onClick={[MockFunction spy]}
    src="/src/icons/skip.svg"
    style={
      {
        "cursor": "pointer",
        "height": 16,
        "width": 16,
      }
    }
    title="Skip forwards"
  />
</div>
`;
